File: 5-Problem-A-Day/147. Insertion Sort List.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        while (head != null) {
            ListNode curr = dummy.next, currPrev = dummy;
            while (true) {
                if (curr == null || curr.val >= head.val) {
                    if (curr != head) {
                        ListNode tmp = head.next;
                        prev.next = tmp;
                        currPrev.next = head;
                        head.next = curr;
                        head = tmp;
                    } else {
                        head = head.next;
                        prev = prev.next;
                    }
                    break;
                } else {
                    currPrev = currPrev.next;
                    curr = curr.next;
                }
            }
        }
        return dummy.next;
    }
}

File: 2021/147. Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* dummy = new ListNode(0);
        while (head != NULL) {
            ListNode *curr = dummy->next, *prev = dummy;
            while (curr != NULL && head->val > curr->val) {
                prev = curr;
                curr = curr->next;                
            }
            prev->next = head;
            head = head->next;
            prev->next->next = curr;
        }
        head = dummy->next;
        delete dummy;
        return head;
    }
};